src/private.rs

```diff
@@ -1,51 +1,62 @@
-use std::collections::VecDeque;
-use std::sync::{Arc, Mutex};
-use std::io::BufWriter;
-use std::fs;
-
-#[derive(Default, Debug)]
-pub struct FsSync<T> {
-    pub receiver_read_id: u64,
-    pub receiver_idx: Option<usize>,
-    pub receiver_max_idx: Option<usize>,
-
-    pub write_bound: Option<usize>,
-    pub writes_to_read: usize,
-
-    pub sender_idx: usize,
-    pub sender_captured_recv_id: u64,
-    pub sender_fp: Option<BufWriter<fs::File>>,
-
-    pub in_memory_idx: usize,
-    pub bytes_written: usize,
-    pub disk_writes_to_read: usize,
-    pub sender_seq_num: usize,
-    pub mem_buffer: VecDeque<T>,
-    pub disk_buffer: VecDeque<T>,
+use sender;
+use deque;
+use std::{cmp, fs, io};
+use std::path::Path;
+
+#[derive(Debug)]
+pub enum Placement<T> {
+    Memory(T),
+    Disk(usize),
 }
 
-impl<T> FsSync<T> {
-    pub fn new(cap: usize) -> FsSync<T> {
-        FsSync {
-            receiver_read_id: 0,
-            receiver_idx: None,
-            receiver_max_idx: None,
+impl<T> Placement<T> {
+    pub fn extract(self) -> Option<T> {
+        match self {
+            Placement::Memory(elem) => Some(elem),
+            Placement::Disk(_) => None,
+        }
+    }
+}
 
-            write_bound: None,
-            writes_to_read: 0,
+pub type Queue<T> = deque::Queue<Placement<T>, sender::SenderSync>;
+
+pub fn read_seq_num(data_dir: &Path) -> io::Result<usize> {
+    let mut max = 0;
+    for directory_entry in fs::read_dir(data_dir)? {
+        let num = directory_entry?
+            .file_name()
+            .to_str()
+            .unwrap()
+            .parse::<usize>()
+            .unwrap();
```

Comment:
Historically we've assumed that if someone fiddles with hopper's queue directory it should just fail to start or otherwise fall over. Maybe that's a bad assumption?